project-koku/koku
# Backend integration with Platform Sources

Addresses #891 

Backend integration with Platform-Sources.  This change includes integration for OCP and AWS.  For the time being the AWS roleARN is stored in the password field of the Platform-Services Authentication secure data store.  Since the Platform-Sources backend does not handle application-specific Source setup a `/billing_source` endpoint was added to the Sources Integration service to allow adding of the S3 bucket for AWS providers.

Items that still need to be addressed are:
1. Add Azure provider creation.
2. Add OpenShift template
3. Get Platform-Sourceses backend running with docker-compose to finish off the docker-compose based deployment.
4. Potentially move from HTTP API based provider creation and deletion to a direct database implementation if necessary.  The reason why I didn't go down this road initially is because we have a lot of logic tied to the /providers endpoint that would need to be handled going the direct database route.  The churn (both in terms of code changes and risk) didn't seem justified at this point in time.

**Testing**

1. Add OCP Source and verify Koku Provider is created.
2. Add AWS Soruce and verify Koku Provider is created.
3. Delete OCP Source verify Koku Provider is removed.
4. Delete AWS Source verify Koku Provider is removed.
5. Stop Koku API server.  Add source, verify that provider is created once koku API is running again.
6. Stop Sources integration service.  Remove source from step 5.  Verify that provider is deleted once service is running again.
7. Add Source when both koku API and Sources integraiton services are down.  Start API and intgration service one by one, and verify provider is created when both are running.
8. With Sources integration service down.  Add and remove the same source.  Start integraiton service and verify that the provider was not created in koku.
9.  Add OCP source and attempt to add a billing_source via the /billing_source API.  Verify that it fails.
10. Add billing_source to a no existent source.  Verify that it fails.
11. Attempt to start Sources integraiton service with Sources backend down.  Verify warning message is displayed.

**Test Results**
[sources_backend_ut.txt](https://github.com/project-koku/koku/files/3611032/sources_backend_ut.txt)
